Userland/watch.cpp

```diff
@@ -0,0 +1,169 @@
+ *
+ * 1. Redistributions of source code must retain the above copyright notice, this
+ *    list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <AK/String.h>
+#include <AK/StringBuilder.h>
+#include <AK/Time.h>
+#include <AK/Vector.h>
+#include <LibCore/ArgsParser.h>
+#include <stdio.h>
+#include <sys/time.h>
+#include <sys/wait.h>
+#include <time.h>
+#include <unistd.h>
+
+static int opt_interval = 2;
+static bool flag_noheader = false;
+static bool flag_beep_on_fail = false;
+static volatile int exit_code = 0;
+static volatile pid_t child_pid = -1;
+
+String build_header_string(const Vector<const char*>& command, const struct timeval& interval)
+{
+    StringBuilder builder;
+    builder.appendf("Every %d", interval.tv_sec);
+    builder.appendf(".%ds: \x1b[1m", interval.tv_usec / 100000);
+    builder.join(' ', command);
+    builder.append("\x1b[0m");
+    return builder.build();
+}
+
+struct timeval get_current_time()
+{
+    struct timespec ts;
+    struct timeval tv;
+    clock_gettime(CLOCK_MONOTONIC, &ts);
+    timespec_to_timeval(ts, tv);
+    return tv;
+}
+
+int64_t usecs_till(const struct timeval& now, const struct timeval& targ)
+{
+    struct timeval diff;
+    timeval_sub(targ, now, diff);
+    return 1000000 * diff.tv_sec + diff.tv_usec;
+}
+
+void handle_signal(int signal)
+{
+    if (child_pid > 0) {
+        int status;
+        kill(child_pid, signal);
+        waitpid(child_pid, &status, 0);
+        if (WEXITSTATUS(status) != 0) {
+            exit_code = 1;
+        }
+    }
+    exit(exit_code);
+}
+
+int run_command(const Vector<const char*>& command)
+{
+    child_pid = fork();
+    if (child_pid < 0) {
+        // We failed to fork, so print an error message and update exit code
+        exit_code = 1;
+        perror("fork");
+        return errno;
+    } else if (child_pid == 0) {
+        // We are in the child process, so close stdin and run the command
+        close(STDIN_FILENO);
```

Comment:
Should it really close stdin?

It's a *very* unexpected thing for a process (even more so, for a command line utility) not to have FD 0 open.

To me, it would make sense to either just keep stdin as is, or set it to `/dev/null`.

Reply:
Hmm, I've tested this with the procps-ng implementation, and it appears that one doesn't modify stdin. I think I'll keep it open then.